### ISSUE TITLE
Impossible to create dataset with non-infered labels 

### ISSUE THREAD

**Issue by frost-is** [tensorflow/tensorflow#57698]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

2.9

### Custom Code

No

### OS Platform and Distribution

Ubuntu 20.04

### Python version

3.10

### CUDA/cuDNN version

CuDNN 8.5.0.96 with CUDA 11.7

### GPU model and memory

Nvidia A100, 40Gb VRAM

### Current Behaviour?

```shell
Hi,

I've been experimenting with the ILSVRC2012 Dataset lately, in order to do some transfer learning. 

I've used the tensorflow.keras.utils.image_dataset_from_directory function to create the training dataset, 
and let it infer the labels (since every class has its dedicated directory). 
However, for the validation dataset, the images are mixed in a single directory, so I need to label them. 
When feeding labels to the aforementioned function, the documentation states 

[The "labels" argument can be set as] Either "inferred" (labels are generated from the directory structure), 
None (no labels), or a list/tuple of integer labels of the same size as the number of image files found in 
the directory. Labels should be sorted according to the alphanumeric order of the image file paths 
(obtained via os.walk(directory) in Python). 

So I fed the function a list containing 50,000 integers for the classes of the val dataset. 
I made sure the list contains 50,000 integers ranging from 1 to 1000. However, when I attempt to create 
the dataset, the function returns "Found 50,000 files belonging to 1 classes." and no other output.

Despite this, my models infer on the dataset seamlessly.

I don't know if the problem stems from the documentation or the function, so I file it as a bug.

I thank you for your time and attention !
```

### Standalone code to reproduce the issue

```shell
path_val = './ILSVRC/Data/val/'
path_val_labels = './ILSVRC/Data/val.txt'
val_labels = list(np.loadtxt(path_val_labels).astype(int))

val_dataset_224 = tf.keras.utils.image_dataset_from_directory(
  path_val, 
  labels=val_labels, 
  image_size=(224, 224),
  batch_size=32, 
  color_mode='rgb', 
  shuffle=True, 
  seed=None, 
  validation_split=None, 
  subset=None,
  interpolation='bilinear', 
  follow_links=False,
  crop_to_aspect_ratio=False)
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@frost-is,
 To expedite the trouble-shooting process, could you please provide a complete code you are using. It helps to analyse the issue. Thank you!

**Comment by frost-is** [comment]
Hi ! 

The relevant code I'm using is the following : 

`import tensorflow as tf
import numpy as np
import os
import cv2
import time
import tensorflow_model_optimization as tfmot
import pandas as pd
import tensorflow_datasets as tfds

print(tf.test.gpu_device_name())

path_val = './ILSVRC/Data/val/'
path_val_labels = './ILSVRC/Data/val.txt'
val_labels = list(np.loadtxt(path_val_labels).astype(int))

val_dataset_224 = tf.keras.utils.image_dataset_from_directory(
  path_val, 
  labels=val_labels, 
  image_size=(224, 224),
  batch_size=32, 
  color_mode='rgb', 
  shuffle=True, 
  seed=None, 
  validation_split=None, 
  subset=None,
  interpolation='bilinear', 
  follow_links=False,
  crop_to_aspect_ratio=False)

mobilenet_small = tf.keras.applications.MobileNetV3Small(
    input_shape=None, alpha=1.0, minimalistic=False, include_top=True,
    weights='imagenet', input_tensor=None, classes=1000, pooling=None,
    dropout_rate=0.2, classifier_activation='softmax',
    include_preprocessing=True
)

output_mobilenetv3_small = mobilenet_small.predict(val_dataset_224))

`

The outputs I get are "found 50000 files belonging to 1 classes", then "1563/1563 [==============================] - 18s 11ms/step", then the array of predictions. The array of predictions is a 50,000 by 1,000 numpy ndarray. 

The "val.txt" file can be found here : https://github.com/fh295/semanticCNN/blob/master/imagenet_labels/ILSVRC2012_validation_ground_truth.txt

The validation dataset is just the validation file of the complete imagenet dataset, and can be found on kaggle https://www.kaggle.com/c/imagenet-object-localization-challenge/data or via academictorrents : https://academictorrents.com/collection/imagenet-2012

Thanks !

**Comment by frost-is** [comment]
Hi again - quick update

I'm rather puzzled - to be completely sure there was a problem with the lables, I've tried checking whether the dataset had labels using `y = np.concatenate([y for x, y in val_dataset_224], axis=0)`, and... It outputs a numpy ndarray of size 50,000 containing integers between 1 and 1000 - so it seems the dataset does have labels. Thinking it could have merely been a problem with the logging of the `image_dataset_from_directory` function, I tried launching a training. I used 

```mobilenet_small.compile(loss='sparse_categorical_crossentropy', optimizer='adam')
monitor = tf.keras.callbacks.EarlyStopping(monitor='val_loss', min_delta=1e-3, patience=5, verbose=2, mode='auto', restore_best_weights=True)
mobilenet_small.fit(train_dataset_224, 
validation_data=val_dataset_224, callbacks=[monitor], verbose=2, batch_size=32, epochs=10)```

And it did start training. But after one epoch, I got 

`Epoch 1/10
40037/40037 - 1457s - loss: 2.0037 - val_loss: nan - 1457s/epoch - 364ms/step`

So it seems the labels are there but unexploitable ? 
At this point I must admit I'm not a sufficiently advanced user to understand where the problem lies 

I hope this helps !

**Comment by tilakrayal** [comment]
@frost-is,
The files which are provided consist of a huge amount of data which we are unable to download and difficult to reproduce the issue. Could you please simplify the data and code or provide the colab gist which helps to analyse the issue. Thank you!

**Comment by frost-is** [comment]
Hi,

Sorry for the delay - here is the colab which contains exactly the code I've been using : https://drive.google.com/file/d/1OcsFWK6f2erhRRmItLLceqcOLwq940NO/view?usp=sharing

There's a new problem that appeared when using colab though : when I use the `tf.keras.utils.image_dataset_from_directory` function, it does not detect the presence of the images and I get the error `Expected the lengths of `labels` to match the number of files in the target directory. len(labels) is 5 while we found 0 files in /content/dataset_cats_dogs/.`, when on my server I get `found 5 files belonging to 1 classes`

Apologies for the added problem - I hope that it helps. Thank you!

**Comment by frost-is** [comment]
Alright - this was actually not a bug but a feature, I used the fix described in https://github.com/keras-team/keras/issues/16205

It's working now - and I do get the `Found 5 files belonging to 1 classes.`

For reference, the dataset is a little something I put together for this issue, containing 5 images - 3 of cats (1 3 5) and 2 of dogs (2 4)

Despite feeding a list to the function, it still doesn't take it into account. However, when I try to extract the labels from the dataset I can do it, albeit in the wrong order, for some reason. If there's anything I can do to clarify the issue let me know - thank you ! 

The colab (still the same) : https://drive.google.com/file/d/1OcsFWK6f2erhRRmItLLceqcOLwq940NO/view?usp=sharing

**Comment by frost-is** [comment]
Hi, 

I opened an issue on keras [here](https://github.com/keras-team/keras/issues/17141), I'll post an update when it's fixed 

Thank you for your help !

**Comment by tilakrayal** [comment]
@frost-is,
Could you please feel free to move this issue to closed status, so that we can track the issue there. Thank you!

**Comment by frost-is** [comment]
Sure, no problem. Thank you for your help !
